File: course_2/module_1/examples/bfs.py
from collections import deque

class Graph:
    def __init__(self):
        # Initialize the graph as an adjacency list.
        self.adjacency_list = {}

    def add_edge(self, u, v):
        """
        Add an undirected edge between nodes u and v.
        If the nodes do not exist in the graph, initialize them.
        """
        self.adjacency_list.setdefault(u, []).append(v)
        self.adjacency_list.setdefault(v, []).append(u)

    def bfs(self, start):
        """
        Perform Breadth-First Search (BFS) starting from the node 'start'.
        Returns a dictionary with the shortest distance from 'start' to all reachable nodes.

        Args:
            start: The starting node for BFS.

        Returns:
            distances: A dictionary where the keys are nodes and values are their distances from 'start'.
        """
        if start not in self.adjacency_list:
            raise ValueError("Start node not found in the graph.")  # Handle invalid starting node.

        explored = {start}  # Set to track explored nodes.
        queue = deque([start])  # Queue for BFS.
        distances = {start: 0}  # Dictionary to store distances from 'start'.

        while queue:
            current = queue.popleft()  # Dequeue the next node.
            for neighbor in self.adjacency_list.get(current, []):  # Explore neighbors.
                if neighbor not in explored:  # If the neighbor hasn't been visited.
                    explored.add(neighbor)  # Mark it as visited.
                    queue.append(neighbor)  # Enqueue the neighbor.
                    distances[neighbor] = distances[current] + 1  # Update its distance.

        return distances

    def connected_components(self):
        """
        Identify all connected components in the graph.
        Returns a list of connected components, where each component is a list of nodes.

        Returns:
            components: A list of lists, where each inner list represents a connected component.
        """
        explored = set()  # Set to track explored nodes.
        components = []  # List to store all connected components.

        for node in self.adjacency_list:  # Iterate through all nodes in the graph.
            if node not in explored:  # Start a new BFS if the node is unexplored.
                component = []  # List to store the current connected component.
                queue = deque([node])  # Queue for BFS.

                while queue:
                    current = queue.popleft()  # Dequeue the next node.
                    if current not in explored:  # If the node hasn't been visited.
                        explored.add(current)  # Mark it as visited.
                        component.append(current)  # Add it to the current component.
                        # Enqueue all unexplored neighbors.
                        queue.extend(
                            neighbor for neighbor in self.adjacency_list.get(current, [])
                            if neighbor not in explored
                        )

                components.append(component)  # Add the current component to the list.

        return components


if __name__ == "__main__":
    # Example usage of the Graph class.
    g = Graph()
    g.add_edge(1, 2)
    g.add_edge(2, 3)
    g.add_edge(4, 5)
    g.add_edge(6, 7)
    g.add_edge(7, 8)
    g.add_edge(8, 6)

    print("Breadth-First Search (BFS) Distances:")
    print(g.bfs(1))  # Perform BFS starting from node 1.

    print("\nConnected Components:")
    print(g.connected_components())  # Find all connected components in the graph.

File: course_2/module_1/examples/bfs.go
package main

import (
	"container/list"
	"fmt"
)

// BFSGraph represents an undirected graph using an adjacency list.
type BFSGraph struct {
	adjacencyList map[int][]int // The adjacency list to store graph edges.
}

// NewBFSGraph creates and initializes a new BFSGraph.
func NewBFSGraph() *BFSGraph {
	return &BFSGraph{
		adjacencyList: make(map[int][]int),
	}
}

// AddEdge adds an undirected edge between nodes u and v.
//
// Args:
//
//	u: The start node of the edge.
//	v: The end node of the edge.
func (g *BFSGraph) AddEdge(u, v int) {
	g.adjacencyList[u] = append(g.adjacencyList[u], v)
	g.adjacencyList[v] = append(g.adjacencyList[v], u)
}

// BFS performs Breadth-First Search starting from a given node.
// It calculates the shortest distance from the start node to all reachable nodes.
//
// Args:
//
//	start: The starting node for the BFS traversal.
//
// Returns:
//
//	A map where keys are nodes and values are their distances from the start node.
//	If the start node does not exist in the graph, returns an error.
func (g *BFSGraph) BFS(start int) (map[int]int, error) {
	if _, exists := g.adjacencyList[start]; !exists {
		return nil, fmt.Errorf("start node %d not found in the graph", start)
	}

	distances := make(map[int]int) // Store distances from the start node.
	visited := make(map[int]bool)  // Track visited nodes.
	queue := list.New()            // Queue for BFS.

	// Initialize BFS.
	queue.PushBack(start)
	visited[start] = true
	distances[start] = 0

	// Perform BFS traversal.
	for queue.Len() > 0 {
		element := queue.Front()
		queue.Remove(element)

		current := element.Value.(int)

		// Explore all neighbors of the current node.
		for _, neighbor := range g.adjacencyList[current] {
			if !visited[neighbor] {
				visited[neighbor] = true
				queue.PushBack(neighbor)
				distances[neighbor] = distances[current] + 1
			}
		}
	}

	return distances, nil
}

// ConnectedComponents identifies all connected components in the graph.
//
// Returns:
//
//	A slice of slices, where each inner slice represents a connected component.
func (g *BFSGraph) ConnectedComponents() [][]int {
	visited := make(map[int]bool)  // Track visited nodes.
	components := make([][]int, 0) // Use `make` to initialize the slice.

	// Explore all nodes in the graph.
	for node := range g.adjacencyList {
		if !visited[node] {
			component := make([]int, 0) // Initialize the component slice.
			queue := list.New()         // Queue for BFS.

			// Initialize BFS for this component.
			queue.PushBack(node)
			visited[node] = true

			// Perform BFS for the current component.
			for queue.Len() > 0 {
				element := queue.Front()
				queue.Remove(element)

				current := element.Value.(int)
				component = append(component, current)

				for _, neighbor := range g.adjacencyList[current] {
					if !visited[neighbor] {
						visited[neighbor] = true
						queue.PushBack(neighbor)
					}
				}
			}

			// Append the component to the result.
			components = append(components, component)
		}
	}

	return components
}

func main() {
	// Create a new graph.
	graph := NewBFSGraph()

	// Add edges to the graph.
	graph.AddEdge(1, 2)
	graph.AddEdge(2, 3)
	graph.AddEdge(4, 5)
	graph.AddEdge(6, 7)
	graph.AddEdge(7, 8)
	graph.AddEdge(8, 6)

	// Perform BFS starting from node 1 and print distances.
	fmt.Println("Breadth-First Search (BFS) Distances:")
	if distances, err := graph.BFS(1); err != nil {
		fmt.Println(err)
	} else {
		// Format distances as {node: distance}.
		fmt.Print("{")
		first := true
		for node, distance := range distances {
			if !first {
				fmt.Print(", ")
			}
			fmt.Printf("%d: %d", node, distance)
			first = false
		}
		fmt.Println("}")
	}

	// Find all connected components in the graph and print them.
	fmt.Println("\nConnected Components:")
	components := graph.ConnectedComponents()
	fmt.Print("[")
	for i, component := range components {
		if i > 0 {
			fmt.Print(", ")
		}
		fmt.Print(component)
	}
	fmt.Println("]")
}
